src/concurrent/mt_scheduler.cr

```diff
@@ -0,0 +1,304 @@
+# this thread.
+#
+# Only the class methods are public and safe to use. Instance methods are
+# protected and must never be called directly.
+class Crystal::Scheduler
+  @@mutex = uninitialized Thread::Mutex
+  @@all_runnables = uninitialized Array(Crystal::Scheduler::Runnables)
+
+  @@park_mutex = uninitialized Thread::Mutex
+  @@park_cond = uninitialized Thread::ConditionVariable
+  @@park_count = uninitialized Int32
+
+  # :nodoc:
+  def self.init
+    @@mutex = Thread::Mutex.new
+    @@all_runnables = Array(Crystal::Scheduler::Runnables).new(Crystal::NPROCS + 1)
+
+    @@park_mutex = Thread::Mutex.new
+    @@park_cond = Thread::ConditionVariable.new
+    @@park_count = 0
+
+    # allocate the main scheduler:
+    Thread.current.scheduler
+  end
+
+  def self.current_fiber : Fiber
+    Thread.current.scheduler.@current
+  end
+
+  def self.enqueue(fiber : Fiber) : Nil
+    Thread.current.scheduler.enqueue(fiber)
+  end
+
+  def self.enqueue(fibers : Enumerable(Fiber)) : Nil
+    Thread.current.scheduler.enqueue(fibers)
+  end
+
+  def self.reschedule : Nil
+    Thread.current.scheduler.reschedule
+  end
+
+  def self.resume(fiber : Fiber) : Nil
+    Thread.current.scheduler.resume(fiber)
+  end
+
+  def self.sleep(time : Time::Span) : Nil
+    Thread.current.scheduler.sleep(time)
+  end
+
+  def self.yield : Nil
+    Thread.current.scheduler.yield
+  end
+
+  def self.yield(fiber : Fiber) : Nil
+    Thread.current.scheduler.yield(fiber)
+  end
+
+  # :nodoc:
+  def initialize(@main : Fiber)
+    @current = @main
+    @runnables = Crystal::Scheduler::Runnables.new
+    @rng = Random::PCG32.new
+    @@mutex.synchronize { @@all_runnables << @runnables }
+  end
+
+  # :nodoc:
+  #
+  # The scheduler's main loop. Follows the algorithms from the "Scheduling
+  # Multithreaded Computations by Work Stealing" (2001) paper.
+  #
+  # The main loop runs on the thread's main stack, and first tries to dequeue
+  # and resume a fiber from its internal runnables queue.
+  #
+  # Once the queue is emptied, the scheduler immediately resumes its main loop,
+  # and becomes a thief that will repeat the following actions until it gets a
+  # runnable fiber that it can resume:
+  #
+  # 1. passes CPU time to another thread, allowing other threads to progress;
+  # 2. tries to steal a fiber from another random scheduler;
+  # 3. executes the event loop to try and fill its internal queue.
+  def start : Nil
+    Thread.log "scheduler#start", self
+
+    # always process the internal runnables queue on startup:
+    if fiber = @runnables.pop?
+      resume(fiber)
+    end
+
+    # the runnables queue is now empty
+
+    while Crystal.running?
+      if fiber = steal_loop
+        resume(fiber)
+
+        # the runnables queue is now empty again
+      else
+        park
+      end
+    end
+  end
+
+  protected def steal_loop : Fiber?
+    Thread.log "scheduler#steal_loop", self
+
+    100.times do
```

Comment:
I am just curious why 100?
This seems like a high number but I do not do much MT programming.

Reply:
Magic number. It should probably be much higher, to avoid parking the thread to quickly, while enqueue should wait for the queue to have some entries before unparking a thread, to make it worth it.

Reply:
Maybe it should be adaptive, too, depending on how much time was spent waiting in the previous steal loop, and whether the thread was parked or not.